### ISSUE TITLE
Reporting status colours and categories

### ISSUE THREAD

**Issue by AnnCorp** [ONSdigital/sdg-indicators#2851]
Currently UK platform has 3 colours/categories:  Exploring data sources (red) Statistics in progress (Amber), Reported online (Green). The US has similar categories plus Improving measures (grey) - think this was tier3? Australia has slight variation - see https://www.sdgdata.gov.au/reporting-status
Feedback from UK team and African partners:
- remove the current RAG colours used for reporting status of data and reserve for assessments of progress towards the indicators/targets/goals (UK currently exploring these types of assessments)
- need a category to show where an indicator is not relevant eg Rwanda commented that as they are land locked some indicators not relevant to them. 
- UK used 'statistics in progress' category a lot initially but is now rarely used - happy to just show if data is in the system or not.
- the reporting status screen doesn't show which indicators are reported and which aren't - just totals - could this evolve
Need to review and agree way forward - this issue to be used to discuss requirements/options
Whichever colours we move to (if we continue to use colours), we'd need to consider accessbility colour contrast testing and how well would work in high contrast version.

 [UPDATE From @dougmet ]

Want to generalise in two ways.

1. Would like to have top level flag along lines of, `willreport`. Will we ever report this statistic
2. For reported statistics generalise reporting status so allowed values can be configured individually for each country. UK will only use two statuses and want to set our own colours. This will be set in a config file on either the data or web sites.

**Comment by AnnCorp** [comment]
Categories: 
Think we clearly need a 'Reported online' and a 'Not yet reported' or 'Not reported' depending how we define.
We need some kind of 'Not applicable' for where indicators not relevant in a country cointext
Need to consider how categorise where methodologies not yet defined so can't report - tier 3 (how many of these are there now?)

**Comment by AnnCorp** [comment]
@brockfanning  one of the things to start to discuss at meeting next Friday please.

**Comment by brockfanning** [comment]
@AnnCorp Just confirming, the grey "Improving measures" in the [US platform](https://sdg.data.gov/status/) does mean a "un_designated_tier" of "3". One note - some tier 3 indicators actually have data in the US platform, so to prevent them from showing up grey, they have been given a "un_designated_tier" of "3 (with data)" - essentially tricking the Jekyll logic. An example is [17-3-1](https://sdg.data.gov/17-13-1/).

Also to rule out this possibility: A quick-fix for the not-applicable problem is to simply remove those indicators from the site. That may not be desirable, but just making sure it's mentioned (since it's easy to do). But assuming you'd prefer a true "not-applicable" category, I'm happy to help work that in. It would either correspond to a new value on an existing metadata field, or to a new metadata field (eg, "indicator_not_applicable").

Regarding the status page showing which specific indicators are in which categories: Just my 2 cents, but I think that would make the status page too busy.

A possible quick fix for the high-contrast version: We could hide the RAG line when in high-contrast mode, which would allow us to make the totals simple high-contrast black-and-white.

**Comment by Norric1Admin** [comment]
'inprogress' option removed. Have begun investigating changing colours.
https://trello.com/c/7gZ5Yxh1/386-remove-in-progress-from-reporting-status-update-colour-scheme

**Comment by dougmet** [comment]
We want to make this much more general for open-sdg. We're thinking of having a couple bits:

1. A `not-applicable` flag (or equivalent name)
2. Custom reporting statuses and colours, specified in the `_config.yml` *or* in the `prose.yml` on the data site, which surfaces as the schema on the main site.

**Comment by dougmet** [comment]
At the moment things are relatively hard-coded on the JavaScript side. I'm currently thinking of something that will generate reporting statistics on the backend (python) and expose the results in some JSON that looks something like (tbc):

`/data/stats/reporting.json`

```json
{  
  "overall":{  
    "complete":{  
      "count":200,
      "percentage":82
    },
    "notstarted":{  
      "count":43,
      "percentage":18
    }
  },
  "goals":{  
    "1":{  
      "complete":{  
        "count":14,
        "percentage":88
      },
      "notstarted":{  
        "count":2,
        "percentage":12
      }
    },
    "2":{  
      "complete":{  
        "count":5,
        "percentage":50
      },
      "notstarted":{  
        "count":5,
        "percentage":50
      }
    }
  }
}
``` 

Then the front end would pick this up and build the whole reporting status page.

The other option is to attempt to build something generic on the front-end. My main reason for favouring the backend is that it's a potentially useful element to the data API.

**Comment by dougmet** [comment]
I think I need a day on the backend and James would need 2 on the front. So estimate is 3-days.

**Comment by dougmet** [comment]
Want to include `notreported` counts in here too @AnnCorp ? It might require another level of nesting but is relatively straight forward:

```json
"2": {
  "totals": {
    "n_total": 12
    "n_applicable": 10
  },
   "statuses":{  
        "complete":{  
          "count":5,
          "percentage":50 # could be computed from `count / n_applicable`
        },
        "notstarted":{  
          "count":5,
          "percentage":50
        }
    }
}
```

**Comment by dougmet** [comment]
If we use something between the two we might even be able to build the page in Jekyll, which is better for the non-JavaScript people

**Comment by Norric1Admin** [comment]
@dougmet 
@AnnCorp will be back tomorrow (around 10:30). Agree that we should design something on the backend and I think that the intermediate option (build in Jekyll) sounds like the most flexible option. In the option you posted above we have 3 categories (4 if you include total): complete. notstarted and N/A. That sounds perfect to me.

Interested to see how this progresses. Will it be possible to include some flexibility for the user, so that categories can be added or edited in future?

**Comment by Norric1Admin** [comment]
For the timings though (3 days) I would need @AnnCorp to confirm.

**Comment by dougmet** [comment]
yes it's all very flexible at this stage. I imagine any number of categories is allowed but they need to be specified somewhere so that we can run data validation as part of the build.

**Comment by dougmet** [comment]
@JamesBedford this is a first effort. I can rearrange it any which way, I thought this might be the most suitable for iterating over.

https://gist.github.com/dougmet/98a7be9b5c92e0ab0bfe6aee141bf6fa

**Comment by dougmet** [comment]
Very open to wording on things like `applicable` vs `count` etc.

**Comment by JamesBedford** [comment]
Hi there!

What’s this for? Have you got the right James Bedford?

Thanks,

James

**Comment by dougmet** [comment]
Whoops, no that was a mistake! Should have been @JABedford - my apologies! 😬

**Comment by Norric1Admin** [comment]
Good progress +1
Ann is happy with the estimate of 3 days.

**Comment by dougmet** [comment]
Based on conversations from recent conference call. Do we need to plan for the possibility that not all indicators are SDG and we might need to filter that? Or is that engineering for problems that don't yet exist?

**Comment by JamesBedford** [comment]
Please remove me from future correspondence.

Thanks! Have a great day!

James

**Comment by dougmet** [comment]
ah, are you being spammed every time we comment. OK. Um, I'll try deleting everything in this thread and then you may need to "unwatch" the issue or something. Apologies again.

**Comment by Norric1Admin** [comment]
@jamesbedford Click ' Unsubscribe' under Notifications and you'll stop receiving updates. You too!

**Comment by dougmet** [comment]
yep apologies. It's this one 
![image](https://user-images.githubusercontent.com/5878305/53101012-cd3c0880-3520-11e9-96ca-8837f6ce4023.png)

I'll clean up the thread in a mo but I can't unsubscribe.

**Comment by Norric1Admin** [comment]
As far as I know, there are no plans to add additional non-SDG indicators to the UK site - this is mainly something being considered by partner countries. It might be useful to have a Boolean 'sdg' field in the metadata, but I wonder if this opens a can of worms - we might redesign this approach anyway in future as we don't know how these would be displayed. @AnnCorp any thoughts?

tldr; good idea but fine to continue without 'sdg' field I think

**Comment by AnnCorp** [comment]
I think it would be something to bear in mind that people may reuse for wider reporting than SDG, but that's not something we would have capacity to include in the development presently and is definitely beyond our SDG remit.

**Comment by dougmet** [comment]
Have hit a minor barrier with translations. This could make it difficult to have fully flexible statuses. It might be more of a case that you can choose from a limited set. So there are three now, we may add something along the lines of `notapplicable` and individual countries can choose fewer (so for example the UK doesn't want `inprogress`.

I propose to do it this way in the first instance, use the reporting built from the data service, and some conditional logic on which categories to show, but think about how a more generic translated version could work. At the moment I'm favouring "choose from this finite set of statuses" where each one has a translation in each country.

**Comment by brockfanning** [comment]
@dougmet Haven't seen your code yet, but it's possible that the "t" filter I added to the jekyll-open-sdg-plugins Gem will help. It's main use-case is translating something that exists as a variable, like from the site config or from indicator metadata. I've got several pull-requests on open-sdg awaiting review, but here is a really simple one to illustrate what I'm talking about: https://github.com/open-sdg/open-sdg/pull/96

In a nutshell: If the string passed into the "t" Liquid filter is a dot-delimited string corresponding to a "translation key" (like from the sdg-translations project) then it will output the translated text. Otherwise it will just output the original string.

Hopefully that's relevant/helpful here?

**Comment by dougmet** [comment]
Yes very relevant and helpful. I'll give this a go.

Work in progress at open-sdg/sdg-build#14

**Comment by dougmet** [comment]
This is what I'm adding to prose: ONSdigital/sdg-data@5d8a696987861e866044fd97c0505f83ef013e54

**Comment by dougmet** [comment]
I'm loving `translation_key | t`, this is really neat.

**Comment by brockfanning** [comment]
Cool! Yeah I'm excited to get some of those PRs into open-sdg - I feel like that precedent opens the door to fixing a lot of nagging multilingual issues.

**Comment by dougmet** [comment]
I've made pretty good progress now. Although as the changes are in multiple places it may take a short while to get everything lined up and merged in. Summary as follows:

## Done
### sdg-build
open-sdg/sdg-build#14 brings new functionality to the data build. This creates a `stats` route, with reporting.json that looks a bit like this:

https://gist.github.com/dougmet/cdc5352cf0db87baf58d661d3b742919

### open-sdg
From that reporting data we can actually build the `reportingstatus.html` page without any JavaScript. So `reportingStatus.js` is removed and  `reportingstatus.html` builds the whole page from `reporting.json`.

See [open-sdg/open-sdg@reporting-status-nojs](https://github.com/open-sdg/open-sdg/tree/reporting-status-nojs) for my current implementation. This will need some changes to line up with all the t-filter changes coming.

### sdg-indicators
I do have one lookup for which I couldn't think of a better way than putting it in the `_config.yml`:
```yaml
status_class:
  notstarted: "danger"
  inprogress: "warning"
  complete: "success"
```
This maps the `reporting_status` to a css class.

The current example is in [ONSdigital/sdg-indicators@2851-reporting](https://github.com/ONSdigital/sdg-indicators/tree/2851-reporting)

### sdg-data

The big change here is that `_prose.yml` now needs some extra bits:

* If a status is not in the options then it's assumed you don't use it
* Add `translation_key` to each `reporting_status` option
* The *ordering* of the reporting statuses matters because this is how it's ordered on the site.

## Not Done
I *haven't* addressed `notapplicable` yet. My first job was to match current functionality and then it might be apparent how best to deal with that status/flag.

**Comment by brockfanning** [comment]
Nice! Question about the CSS classes - do you think it would be feasible to "slugify" a status' translation key to create it's CSS class? We might have to add some extra selectors to the open-sdg CSS so that it works out-of-the-box, but I think that's fine.

**Comment by dougmet** [comment]
Yes I think that would be nice. I've been trying to keep it working as I go (hence the lookup) but a better solution would likely be exactly that. On the CSS side I only paused because we currently use quite specialised bootstrap classes (`success`, `danger` etc) so would need to think about how to do our own. Could they just inherit from the bootstrap ones and change colour? @JABedford ?

**Comment by JABedford** [comment]
Having a quick look it is just the colour that is defined in that class, it doesn't rely on any other Bootstrap styling so we have a number of different options.

**Comment by dougmet** [comment]
OK so the css part was easy. I took @brockfanning 's suggestion and replaced the success/warning/danger classes with slugified reporting status. The only change in sdg-indicators now is the custom css if you want it.

I also updated the python package with default translations so that people wouldn't get caught out after the upgrade. So now the biggest pain for upgrading is that you must update **sdg-data** to use the latest version of **sdg-build**. Otherwise the build will fail.

I've got this working on my four test branches. https://sdg.mango-solutions.com/2851-reporting/reporting-status/

**Comment by Norric1Admin** [comment]
@dougmet Now that https://github.com/open-sdg/sdg-translations/pull/72 has been merged, is it possible to add the `notapplicable` status (and can it be part of opensdg@v.0.5.0, or will it take more time)?

**Comment by dougmet** [comment]
It will be part of 0.5.0 now by default without anything else needed. If you show me where the change is needed I can send a PR towards the data site (it's a change in the _prose.yml in the country's data site that's required).

**Comment by dougmet** [comment]
And their CSS. I will make some documentation.

**Comment by Norric1Admin** [comment]
Thanks @dougmet - it is for the UK and Rwandan sites:
https://github.com/ONSdigital/sdg-indicators / sdg-data
https://github.com/sustainabledevelopment-rwanda/sdg-indicators / sdg-data

**Comment by AnnCorp** [comment]
Closing this as Open SDG now allows configuration of Reporting status colours, categories and how view (by goals or other criteria)
